Move back to BM25 similarity

javanna opened the pull request:
With the last lucene upgrade we have temporarily adopted the `LegacyBM25Similarity` which exposes the same scores as BM25Similarity before the k1+1 factor was removed from the numerator of the scoring formula.

This PR changes the default Elasticsearch similarity back to the `Bm25Similarity` and updates the scores that have changed due to such change in our docs and tests.

javanna committed:
Move back to BM25 similarity

With the recent lucene upgrade we have temporarily adopted the LegacyBM25Similarity which exposes the same scores as BM25Similarity before the k1+1 factor was removed from the numerator of the scoring formula.

This commit moves the default Elasticsearch similarity back to the Bm25Similarity and updates the scores that have changed in our docs and tests

javanna committed:
fix tests

javanna committed:
add note about bm25 changes

javanna committed:
update scores in docs

javanna committed:
update scores in docs

javanna committed:
fix bad merge

javanna committed:
fix sql tests

elasticmachine commented:
Pinging @elastic/es-search

romseygeek commented on server/src/test/java/org/elasticsearch/search/nested/SimpleNestedIT.java:
```diff
@@ -326,7 +326,7 @@ public void testExplain() throws Exception {
         Explanation explanation = searchResponse.getHits().getHits()[0].getExplanation();
         assertThat(explanation.getValue(), equalTo(searchResponse.getHits().getHits()[0].getScore()));
-        assertThat(explanation.toString(), startsWith("0.36464313 = Score based on 2 child docs in range from 0 to 1"));
+        assertThat(explanation.toString(), startsWith(explanation.getValue() + " = Score based on 2 child docs in range from 0 to 1"));
```
👍

romseygeek commented on x-pack/plugin/sql/qa/src/main/resources/docs.csv-spec:
```diff
@@ -713,8 +713,8 @@ SELECT SCORE(), * FROM library WHERE MATCH(name, 'dune') ORDER BY SCORE() DESC;
-1.8893257      |Frank Herbert  |Dune Messiah       |331            |1969-10-15T00:00:00Z
+1.0402887       |Frank Herbert  |Dune               |604            |1965-06-01T00:00:00Z
+0.8587844      |Frank Herbert  |Dune Messiah       |331            |1969-10-15T00:00:00Z
 1.6086555      |Frank Herbert  |Children of Dune   |408            |1976-04-21T00:00:00Z
```
This doesn't look right?  All the scores should change, and currently things aren't in score order

javanna replied:
ok I will double check if I fixed all the issues in tests, and if so I will check why this is the case here.

javanna replied:
Let me add that I have just indexed the same documents used for these tests and verified manually that all of the 4 documents matching this query get an updated score with the bm25 change. Seems like something weird happens with these tests, not sure what exactly.

jimczi replied:
A mixed cluster with nodes in 6x and 7 ?

javanna replied:
possibly, yet I would expect these tests to run both in multi-node and single-node. which makes me expect that I should not get a green build with the current changes. Single-node should require to update all the scores. I may not be getting how test tests are run.

astefan replied:
@javanna indeed, there is something wrong here. The tests pass and, looking at the logs it produces, it seems like it's not comparing the output from the `csv-spec` with the actual output of the query being executed... am looking into it.

astefan replied:
@javanna the issue is [here](https://github.com/elastic/elasticsearch/blob/6fbc166afd75776363ebba897fcf032ce8a9d999/x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcAssert.java#L223). An assertion with 1.0 delta which, in the case of SCORE(), is quite relevant and a deal breaker.

romseygeek reviewed (commented):
One question around the SQL tests, looks good otherwise.

javanna committed:
Merge branch 'master' into enhancement/back_to_bm25

javanna committed:
fix remaining failures

javanna committed:
revert change sto release-notes.asciidoc

javanna committed:
Merge branch 'master' into enhancement/back_to_bm25

romseygeek reviewed (approved):
LGTM, thanks @javanna!

javanna committed:
update release notes - mention mixed clusters

javanna committed:
Merge branch 'master' into enhancement/back_to_bm25

javanna committed:
Add legacy bm25 to default and built-in similarities

This allows users to opt-out of the updated bm25 similarities and use
the deprecated legacy bm25 similarity.

This helps especially cross-cluster search cases across clusters on
multiple versions, otherwise sorting by score would lead to very weird
results depending on the version of the data node that scores each doc.

javanna committed:
additional test

javanna committed:
Merge branch 'master' into enhancement/back_to_bm25

javanna commented:
retest this please

javanna commented:
run gradle build tests 1

javanna committed:
Merge branch 'master' into enhancement/back_to_bm25

javanna committed:
use legacy bm25 for indices created in 6.x

javanna commented:
run gradle build tests 2

javanna committed:
add rolling upgrade test

javanna committed:
add integration test

javanna commented:
I pushed an update around backwards compatibility: we now use legacy bm25 for indices created in 6.x, instead of the updated bm25. That makes sure that in a mixed cluster, scores for the same documents don't change depending on whether their shard is located on 7.0 or 6.x nodes. That also preserves document ordering when sorting by score against indices created on 6.x.

Indices created once all nodes are on 7.0 though will use the updated bm25 similarity. That is necessary so we expose the updated behaviour but it has a consequence: when searching against both indices created on 6.x and indices created on 7.0, scores will not be comparable (the removal of the constant factor from the numerator causes scores in 7.0 to be much lower compared to how the same documents were scored in 6.x). If this is a problem, users can address it by explicitly using LegacyBM25 for newer indices (created after the upgrade to 7.0 is completed). 

A couple of points left open for discussion:

- shall we make the legacy bm25 similarity available in 6.7 so users can prepare for the migration? I tend to think that this is not necessary as the legacy similarity will anyway be silently applied to 6.x indices, hence users will only need it once all nodes are on 7.0 which is what this PR does.
-  silently using legacy bm25 for 6.x indices means that there is no way for users to force the real bm25 similarity on such indices. Maybe that's ok but I wanted to raise it to make sure. Ideally we would at some point switch the 6.x indices to ordinary BM25 like newer indices rather than having user set legacy bm25 manually for the newer indices, but that is a bigger challenge that we are not ready for at the moment.
- do we want to deprecate legacy bm25 straight-away, do it in later 7.x releases, or leave it around (and deprecate it) in 8.x too as an alias of bm25?

@jimczi would you like to have a look?

javanna commented:
run default distro tests

javanna committed:
improve test

jimczi reviewed (commented):
> silently using legacy bm25 for 6.x indices means that there is no way for users to force the real bm25 similarity on such indices. Maybe that's ok but I wanted to raise it to make sure. Ideally we would at some point switch the 6.x indices to ordinary BM25 like newer indices rather than having user set legacy bm25 manually for the newer indices, but that is a bigger challenge that we are not ready for at the moment.

Yes, this is not ideal. We should make the legacy BM25 available so that users that upgrade to 7 can opt-out from the old formula. Another solution is to use a cluster setting that switches from one impl to the other but this would add a manual operation to the upgrade. My preference goes to a renaming of BM25 to something like BM25_legacy and that indices created in 6x are automatically upgraded to this similarity if they were using BM25. 

> do we want to deprecate legacy bm25 straight-away, do it in later 7.x releases, or leave it around (and deprecate it) in 8.x too as an alias of bm25?

If the legacy BM25 is exposed I'd vote for a deprecation in 7.0 so that users know that we automatically upgraded to a deprecated formula.

s1monw commented:
> @javanna and I discussed this issue a bit, and I want to propose another alternative. We will have 3 similarity options: BM25_6x (if we can find something to replace 6x with - like K1Factor, it will be better), BM25_7x and BM25. The first two are explicit about what they do and the last one maps to the 6.x or the 7.x version based on the creation version of the index. This means that:

can we instead allow a similarity to accept a `compatiblity_version` that is by default the `index.version.created` and can be set by the user. It can be at most `N-1` majors from the `index.version.created` away and we can reuse it in the future if we need it.

javanna commented:
We have re-discussed this change with @jimczi , @jpountz and @s1monw . We were not too happy with the migration path and the possible solutions to address the breaking nature of this change. We decided to keep using `LegacyBm25` similarity for the whole 7.x series and work on the migration to the updated bm25 scoring formula for 8.0. Meanwhile, we hope that we will have more tools to deal with such changes, like response versioning. I just marked this PR stalled waiting for #11184 and a blocker for 8.0.

ywelsch commented:
I've removed the blocker label here, as this change will need some time to bake, and won't be ready for 8.0.

javanna commented:
We have re-discussed this with the team. We said we are not incredibly sure how to move away from the legacy bm25 without breaking backwards compatibility, and whether it's even worth the effort. I spent some time thinking about the problem and possible solutions and I see two options:

1) rename legacy bm25 to something more appealing, un-deprecate it, and embrace that we won't easily move away from it. It is a small class and uses BM25Similarity from Lucene besides when it modifies the scoring formula.

2) given that the difference between bm25 and its legacy variant is "only" around how scores are computed during search, would it be possible to default to the newer bm25, and fall back to the legacy one during search whenever at least one node that we are talking to is on an older version (and does not default to it)? Would that be enough to guarantee proper sorting by score hence backwards compatibility both in a CCS scenario as well as in a rolling upgrade situation?

This second option is along the lines of something that we have recently done to emulate the fields option when talking to older nodes that don't support it yet. The coordinating node would need to detect that at least one older node is involved in the search, and ask all of the newer involved nodes to score in a compatible way, by using the legacy bm25 similarity for that specific search. This does sound quite an involved task: it is worth discussing and evaluating, but we may still decide to resort to option 1 after all. Also, there may be other options that I did not consider, but I believe none of the options mentioned above in previous comments are now feasible as they either cause breakage and/or require manual intervention. Thoughts @jpountz ?

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Pinging @elastic/es-search-foundations (Team:Search Foundations)
